### PR TITLE
Including perl side of BRASS

### PR DESCRIPTION
**Pull request by brainstorm**
* [X] I have read the guidelines above.
* [ ] This PR adds a new recipe.
* [X] This PR updates an existing recipe.
* [ ] This PR does something else (explain below).

This is a continuation on PR #2329

### CODE REVIEW

File: recipes/brass/meta.yaml [hunk 1]
```diff
@@ -21,6 +21,12 @@ requirements:
     - perl-grass
     - cansam
     - boost
```
Comment (bgruening) [inline recipes/brass/meta.yaml]: You need to bump the build number.

File: recipes/perl-grass/meta.yaml [hunk 2]
```diff
@@ -25,6 +25,7 @@ requirements:
     - perl-encode-locale
     - perl-lwp-simple
     - perl-xml-parser
```
Comment (bgruening) [inline recipes/perl-grass/meta.yaml]: build-number bump :)

**Comment by brainstorm** [comment]
@bgruening Just bumping the versions it went from green:

https://travis-ci.org/bioconda/bioconda-recipes/builds/178145192

To red?:

https://travis-ci.org/bioconda/bioconda-recipes/builds/178215663

Bizarre :_/

**Comment by bgruening** [comment]
@brainstorm it's not that surprising actually. Without increase the version number Conda is not building your package and not really testing it. Sorry.

**Comment by bgruening** [comment]
Seems you are missing at least: "Can't locate Tabix.pm" as dependency.

**Comment by brainstorm** [comment]
Yes, I identified the problem(s) since, thanks @bgruening... preparing the pkg with an updated release/version of grass now ;)

**Comment by bgruening** [comment]
Great thanks!

**Comment by brainstorm** [comment]
I'm almost there for the `grass` side of dependencies, getting stuck on this error during the build though:

```
(...)
"/anaconda/envs/_build/bin/perl" -MExtUtils::MY -e 'MY->fixin(shift)' -- blib/script/brass_bedpe2vcf.pl
cp bin/grass.pl blib/script/grass.pl
"/anaconda/envs/_build/bin/perl" -MExtUtils::MY -e 'MY->fixin(shift)' -- blib/script/grass.pl
Manifying 14 pod documents
PERL_DL_NONLAZY=1 "/anaconda/envs/_build/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/1_pm_compile.t ............ 1/? Bailout called.  Further testing stopped:  Unable to 'use' module Sanger::CGP::Grass::GenomeData

#   Failed test 'use Sanger::CGP::Grass::GenomeData;'
#   at t/1_pm_compile.t line 56.
#     Tried to use 'Sanger::CGP::Grass::GenomeData'.
#     Error:  Can't load '/anaconda/envs/_build/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/auto/Bio/DB/HTS/HTS.so' for module Bio::DB::HTS: /lib64/libc.so.6: version `GLIBC_2.7' not found (required by /anaconda/envs/_build/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/auto/Bio/DB/HTS/../../../../../../../../libhts.so.1) at /anaconda/envs/_build/lib/perl5/5.22.0/x86_64-linux-thread-multi/DynaLoader.pm line 193.
#  at /anaconda/envs/_build/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/Bio/DB/HTS/Tabix.pm line 22.
# Compilation failed in require at /anaconda/envs/_build/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/Bio/DB/HTS/Tabix.pm line 22.
# BEGIN failed--compilation aborted at /anaconda/envs/_build/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/Bio/DB/HTS/Tabix.pm line 22.
# Compilation failed in require at /anaconda/conda-bld/work/grass-2.1.0/blib/lib/Sanger/CGP/Grass/GenomeData/GenomeDataCache.pm line 27.
# BEGIN failed--compilation aborted at /anaconda/conda-bld/work/grass-2.1.0/blib/lib/Sanger/CGP/Grass/GenomeData/GenomeDataCache.pm line 27.
# Compilation failed in require at /anaconda/conda-bld/work/grass-2.1.0/blib/lib/Sanger/CGP/Grass/GenomeData.pm line 28.
# BEGIN failed--compilation aborted at /anaconda/conda-bld/work/grass-2.1.0/blib/lib/Sanger/CGP/Grass/GenomeData.pm line 28.
# Compilation failed in require at t/1_pm_compile.t line 56.
# BEGIN failed--compilation aborted at t/1_pm_compile.t line 56.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 255 just after 2.
FAILED--Further testing stopped: Unable to 'use' module Sanger::CGP::Grass::GenomeData
make: *** [test_dynamic] Error 255
Command failed: /bin/bash -x -e /bioconda/recipes/perl-grass/build.sh
```

Any ideas on how to solve that `perl-bio-hts` Tabix/ issue are really welcome:

```
#     Tried to use 'Sanger::CGP::Grass::GenomeData'.
#     Error:  Can't load '/anaconda/envs/_build/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/auto/Bio/DB/HTS/HTS.so' for module Bio::DB::HTS: /lib64/libc.so.6: version `GLIBC_2.7' not found (required by /anaconda/envs/_build/lib/perl5/site_perl/5.22.0/x86_64-linux-thread-multi/auto/Bio/DB/HTS/../../../../../../../../libhts.so.1) at /anaconda/envs/_build/lib/perl5/5.22.0/x86_64-linux-thread-multi/DynaLoader.pm line 193.
```

**Comment by bgruening** [comment]
It seems that `Sanger::CGP::Grass::GenomeData` was build on a newer system and is not compatible with our libc.

**Comment by brainstorm** [comment]
I'm confused about this last travisCI error:

https://travis-ci.org/bioconda/bioconda-recipes/jobs/178888322#L1041

Isn't `Bio::Seqfeature::Lite` part of BioPerl already?: https://metacpan.org/pod/Bio::SeqFeature::Lite

**Comment by bgruening** [comment]
Can you try to pin `perl-bioperl-core` to version `1.6.924`

**Comment by bgruening** [comment]
Capture/Tiny.pm is still missing :)

**Comment by brainstorm** [comment]
Thanks for this Bjorn, I have problems reproducing those errors on our cluster and local docker... Thanks for your patience!

**Comment by brainstorm** [comment]
All deps seem to be in place now... Looks like further [mangling of `PERL5LIB` is needed in order for this recipe to work](https://github.com/cancerit/BRASS/blob/dev/setup.sh#L136).

**Comment by bgruening** [comment]
Sorry @brainstorm Sanger/CGP/Brass/Implement.pm is still missing. I feel your pain with perl, really sorry.

**Comment by brainstorm** [comment]
Thanks for all your help @bgruening ;)

It now seems that `Sanger::CGP::Brass::Implement.pm` is [within the BRASS package itself](https://github.com/cancerit/BRASS/blob/dev/perl/lib/Sanger/CGP/Brass/Implement.pm), so not sure why it is not being installed and found in Perl's `@INC` like the other modules...

Perhaps @aferrer and/or @alexm can shed some light on this poorly packaged bioinformatics Perl module? :_)

**Comment by alexm** [comment]
@brainstorm there are several issues that were discussed here, so a bit of
context would make it easier to help. What did you try and what did fail?
How can I try it? Which particular build log should I review?

**Comment by brainstorm** [comment]
@alexm Thanks a ton for showing up here! :)

The offending build log is [over here on TravisCI](https://travis-ci.org/bioconda/bioconda-recipes/jobs/179516655#L1364). Thanks to [bioconda local docker build setup](https://bioconda.github.io/bioconda-utils/contributing.html), which you could use to reproduce this, I've installed all perl dependencies that [BRASS-perl](https://github.com/cancerit/BRASS/tree/dev/perl) requires [via this `meta.yml`](https://github.com/bioconda/bioconda-recipes/pull/3034/files#diff-2ee94adb9124b9c8882c8b42372a070f), but I suspect that the [setup.sh](https://github.com/cancerit/BRASS/blob/dev/setup.sh) and its [mangling of PERL5LIB](https://github.com/cancerit/BRASS/blob/dev/setup.sh#L136) isn't the cleanest install method and is preventing me to install BRASS-perl successfully.

I've managed to [patch and install the C++ version successfully via my recipe](https://github.com/bioconda/bioconda-recipes/tree/master/recipes/brass), but the perl side with its `Sanger::CGP::Brass::Implement module` error (this module part of [BRASS-perl](https://github.com/cancerit/BRASS/blob/dev/perl/lib/Sanger/CGP/Brass/Implement.pm) itself, not 3rd party like the others in `meta.yml`), resists my attempts to install the package.

Thanks again in advance for looking into it!

**Comment by alexm** [comment]
Well, I managed to setup the build environment on a Vagrant box and found that `Sanger/CGP/Brass/Implement.pm` **is not** installed in `/opt/conda/conda-bld/_t_env/bin/../lib` but in `/opt/conda/conda-bld/_t_env/bin/../lib/perl5`.

The `lib` directory is added in `@INC` by this line in `brass.pl` (and many other scripts):

```
use lib "$FindBin::Bin/../lib";
```

Thus, appending `/perl5` to that value or setting PERL5LIB accordingly should do the trick.

**Comment by brainstorm** [comment]
Thanks much @alexm for taking a peek at that!

I bet there are cleaner ways to fix that in Perl, but if I understood what you correctly, something like this `sed` line would do the trick?

```
[rvalls@node08 brass]$ git diff
diff --git a/recipes/brass/build.sh b/recipes/brass/build.sh
index 0c8b20c..bf78e28 100644
--- a/recipes/brass/build.sh
+++ b/recipes/brass/build.sh
@@ -13,6 +13,8 @@ cp c++/filterout-bam $PREFIX/bin
 cd perl && perl -MExtUtils::MakeMaker -e 1
 cpanm --mirror http://cpan.metacpan.org --notest -l $PREFIX --installdeps .

+sed -e 's#$FindBin::Bin/../lib#$FindBin::Bin/../lib/perl5#' bin/*.pl
+
 perl Makefile.PL INSTALL_BASE=$PREFIX && \
 make && \                                                                                                                                                                                                       make test && \
```

Thanks again!

**Comment by alexm** [comment]
It should, but if you're going to edit all those files anyway, then just
get rid of the whole `use lib ...` and set PERL5LIB instead ;-)

**Comment by brainstorm** [comment]
Thought about it but I'm not entirely sure if mangling `PERL5LIB` would play well with the rest of `bioconda`'s ecosystem/packages :_S

**Comment by alexm** [comment]
You can run it with `perl -I/path/to/lib/perl5 brass.pl` too.

**Comment by brainstorm** [comment]
Hm, thanks, neat trick that definitely helps for testing, but the final/installed `brass.pl` should be invoked "as usual" by the user, imho :-S

**Comment by brainstorm** [comment]
Aaand, it actually worked! Only locally though:

```
+ brass.pl --help
Usage:
    brass.pl [options]

      Required parameters:
        -outdir    -o   Folder to output result to.
        -tumour    -t   Tumour BAM file
        -normal    -n   Normal BAM file
        -depth     -d   Regions of excessive sequencing depth to be ignored
        -genome    -g   Genome fasta file
        -species   -s   Species name
        -assembly  -as  Assembly name
        -protocol  -pr  Sequencing protocol (WGS|WXS|RNA)
        -g_cache   -gc  Genome cache file.
        -viral     -vi  Virus sequences from NCBI
        -microbe   -mi  Microbe sequences file prefix from NCBI, please exclude '.N.fa.2bit'
        -gcbins    -b   5 column bed coord file, col 4 number of non-N bases, col 5 GC fraction.
        -centtel   -ct  TSV file of usable regions of the chromosome
                          Example in perl/share/Rdefault/
```

On TravisCI it fails with an (I hope transient) error about fetching info from conda with python's `requests`:

https://travis-ci.org/bioconda/bioconda-recipes/jobs/180386945#L531

@bgruening Bioconda's TravisCI has been acting up since yesterday, any idea of what those conda fetch errors are coming from?

Thanks again @alexm for your touch in this issue and @bgruening for your patience, very appreciated.

**Comment by bgruening** [comment]
@brainstorm  the travis error is due to the build-infrastructure. Please wait a few days until it is fixed! Thanks for all your work.

**Comment by daler** [comment]
@brainstorm build infrastructure problems should be fixed (#3080); can you rebase and retry please?

**Comment by brainstorm** [comment]
Thanks @daler and @bgruening, I did the rebase although something still seems off with bioconda's TravisCI:

https://travis-ci.org/bioconda/bioconda-recipes/jobs/180648103#L369

**Comment by daler** [comment]
Hmm, looks like the rebase didn't quite work since that [url is not used any more in the master branch](https://github.com/bioconda/bioconda-recipes/blob/master/simulate-travis.py#L111).

Could you give it another try, or maybe open up a new pull request?

**Pull request merged**